OSSM-8958: Cluster-wide migration using istio injection label (#89324)
openshift/openshift-docs: modules/ossm-migrating-a-cluster-wide-deployment-using-the-istio-injection-label.adoc

// Module included in the following assemblies:
//

> // * service-mesh-docs-main/migrating/checklists/ossm-migrating-cluster-wide-assembly.adoc

Author: Done

:_mod-docs-content-type: PROCEDURE
[id="ossm-migrating-a-cluster-wide-deployment-using-the-istio-injection-label_{context}"]
= Migrating a cluster-wide deployment by using the istio injection label

You can perform a canary upgrade with the gradual migration of data plane namespaces for a cluster-wide deployment by using the `istio-injection=enabled` label and the `default` revision tag. 

You must re-label all of the data plane namespaces. However, it is safe to restart any of the workloads at any point during the migration process. 

The `bookinfo` application is used as an example for the `Istio` resource. For more information on configuration differences between the {SMProduct} 2 `ServiceMeshControlPlane` resource and the {SMProduct} 3 `Istio` resource, see "ServiceMeshControlPlane resource to Istio resource fields mapping".

> You can follow these same steps with your own workloads.

Author: Done

.Prerequisites

* You have deployed {ocp-product-title} 4.14 or later.
* You are logged in to the {ocp-product-title} web console as a user with the `cluster-admin` role.
* You have completed the premigration checklists.
* You have the {SMProduct} {SMv2Version} Operator installed.
* You have the {SMProduct} 3 Operator installed.

> * You created an `IstioCNI` resource.

Author: Done

> * You have the `istioctl` tool installed.

Author: Done

* You are running a cluster-wide Service Mesh control plane resource.
* You have installed the `bookinfo` application.

.Procedure

. Identify the namespaces that contain a 2.6 control plane by running the following command:
+
[source,terminal]
----
$ oc get smcp -A
----
+
.Example output
[source,terminal]
----
NAMESPACE      NAME                   READY   STATUS            PROFILES      VERSION   AGE
istio-system   install-istio-system   6/6     ComponentsReady   ["default"]   2.6.6     115m
----

> . Create a YAML file named `ossm-3.yaml` that creates the {istio} resource for the 3.0 installation in the same namespace as the `ServiceMeshControlPlane` resource for the 2.6 installation:

Author: Done

+

> [NOTE]

Author: Done

> [NOTE]

Author: For now, I will leave this as a note. We have four related procedures, some of which have already been merged, and they call this out as a note. I'd rather have them all be the same as much as possible. I can discuss with the team after the Monday release.

====

> In the following example configuration, the {istio} control plane has access to all namespaces on the cluster. If you want limit the namespaces the control plan has access to, you must define discovery selectors. All data plane namespaces that you plan to migrate from version 2.6 must be matched.

Reviewer: https://www.ibm.com/docs/en/ibm-style?topic=grammar-verbs#voice

> In the following example configuration, the {istio} control plane has access to all namespaces on the cluster. If you want limit the namespaces the control plan has access to, you must define discovery selectors. All data plane namespaces that you plan to migrate from version 2.6 must be matched.

Author: Done

====
+
.Example `Istio` resource
[source,yaml,subs="attributes,verbatim"]
----

> apiVersion: sailoperator.io/v1alpha1

Reviewer 2: ```suggestion
apiVersion: sailoperator.io/v1
```

Author: Done

kind: Istio
metadata:

>   name: ossm-3 <1>

Author: Done

spec:
  updateStrategy:
    type: RevisionBased

>   namespace: istio-system <2>

Author: Done

  version: v1.24.3
  values:  
    meshConfig:

>       extensionProviders: <3>

Author: Done

        - name: prometheus
          prometheus: {}
        - name: otel
          opentelemetry:
            port: 4317
            service: otel-collector.opentelemetrycollector-3.svc.cluster.local
----
<1> The `name`, `updateStrategy` and `version` fields specify how the `IstioRevision` resource name is created. For more information, see "Identifying the revision name".
<2> The 3.0 and 2.6 control planes must run in the same namespace.
<3> Optional: If you are migrating metrics and tracing, update the `extensionProviders` fields according to your tracing and metrics configurations.
+

> [NOTE]

Author: Will leave as note so this is the same across the set of the related procedures.

// Module included in the following assemblies:
//
// * service-mesh-docs-main/migrating/checklists/ossm-migrating-cluster-wide-assembly.adoc

:_mod-docs-content-type: PROCEDURE
[id="ossm-migrating-a-cluster-wide-deployment-using-the-istio-injection-label_{context}"]
= Migrating a cluster-wide deployment using the istio injection label

> This procedure performs a canary upgrade with the gradual migration of data plane namespaces for a cluster-wide deployment using the `istio-injection=enabled` label. It is safe to restart any workloads at any point during the {SMProduct} {SMv2Version} to {SMProduct} 3 migration process. You must apply new namespace labels.  

Reviewer 2: ```suggestion
This procedure performs a canary upgrade with the gradual migration of data plane namespaces for a cluster-wide deployment using the `istio-injection=enabled` label and `default` revision tag. It is safe to restart any workloads at any point during the {SMProduct} {SMv2Version} to {SMProduct} 3 migration process. You must apply new namespace labels.  
```

Reviewer 2: `It is safe to restart any workloads at any point during the {SMProduct} {SMv2Version} to {SMProduct} 3 migration process. You must apply new namespace labels. `

It does not make much sense now without the context. Original says:
`Compared to the simple Migration of 2.6 installation with istio-injection=enabled label below, it is safe to restart any workloads at any given time during the migration. To achieve this, it requires more manual steps and re-labeling.`

So we are saying that advantage of this approached is that you can restart the workloads at any given time which you can't do in the simple procedure. And also we say that because of this advantage, it's necessary to do annoying namespace relabeling which you don't need to do in the simple approach. We should keep this info in for better understanding.

+
[source,terminal]
----
$ oc get smcp -A
----
+
.Sample output:
[source,terminal]
----
NAMESPACE      NAME                   READY   STATUS            PROFILES      VERSION   AGE
istio-system   install-istio-system   6/6     ComponentsReady   ["default"]   2.6.6     115m
----

. Create a YAML file named `ossm-3.yaml` that creates the {istio} resource for the 3.0 installation in the same namespace as the `ServiceMeshControlPlane` resource for the 2.6 installation.

.Example configuration
[source,yaml,subs="attributes,verbatim"]
----
apiVersion: sailoperator.io/v1alpha1
kind: Istio
metadata:
  name: ossm-3 <1>
spec:
  updateStrategy:
    type: RevisionBased
  namespace: istio-system <2>
  version: v1.24.3
  values:  <3>
    meshConfig:
    extensionProviders:
      - name: prometheus
        prometheus: {}
      - name: otel
        opentelemetry:
          port: 4317
          service: otel-collector.opentelemetrycollector-3.svc.cluster.local
----
<1> The `name`, `updateStrategy` and `version` fields are significant for injection labels.
<2> The 3.0 and 2.6 control planes must run in the same namespace.
<3> Tracing and metrics configurations are optional.
+

> This example configuration does not use any `discoverySelectors`, which means the {istio} control plane has access to all namespaces. If you want to define `discoverySelectors` all data plane namespaces that you plan to migrate from 2.6 must be matched.

Reviewer 2: Indent is broken in the preview somewhere around this line.

Author: done

+
[NOTE]
====

> To prevent the {SMProduct} 3.0 control pane from injecting proxies to workloads in namespaces with istio-injection=enabled label applied do not use use the default name or default revision tag.

Reviewer 2: ```suggestion
To prevent the {SMProduct} 3.0 control pane from injecting proxies to workloads in namespaces with istio-injection=enabled label applied which are still managed by {SMProduct} 2.6 control plan do not use use the `default` name for Istio resource and do not create `default` revision tag at this stage. `default` revision tag  will be created later in this procedure.
```

Author: done

// Module included in the following assemblies:
//
// * service-mesh-docs-main/migrating/checklists/ossm-migrating-cluster-wide-assembly.adoc

:_mod-docs-content-type: PROCEDURE

> [id="ossm-migrating-a-cluster-wide-deployment-using-the-istio-injection-label_{context}"]

Reviewer 3: As per the guideline, you may consider to trim the file name https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#assemblymodule-file-names

Author: There are three different ways to migrate a cluster-wide deployment, so I think I need the lengthy name to make the differences clear.

> = Migrating a cluster-wide deployment using the istio injection label

Reviewer 3: ```suggestion
= Migrating a cluster-wide deployment by using the istio injection label
```

Author: done

> This procedure performs a canary upgrade with the gradual migration of data plane namespaces for a cluster-wide deployment using the `istio-injection=enabled` label and the `default` revision tag. 

Author: done

> This procedure contains more steps than the simple migration method, and you must re-label all of the data plane namespaces. However, while performing this procedure, it is safe to restart any of the workloads at any point during the migration process. 

Author: done

> This procedure contains more steps than the simple migration method, and you must re-label all of the data plane namespaces. However, during the {SMProduct} {SMv2Version} to {SMProduct} 3 migration process using the `istio-injection=enabled` label, it is safe to restart any of the workloads at any point during the migration process. 

Author: done

The `bookinfo` example application is being used for demonstration purposes with a minimal example for the `Istio` resource. For more information on configuration differences between the {SMProduct} 2 `ServiceMeshControlPlane` resource and the {SMProduct} 3 `Istio` resource, see "ServiceMeshControlPlane resource to Istio resource fields mapping".

> You can follow these same steps with your own workloads.

Author: This sentence is in other tasks: some for cluster-wide migration and some for multitenant. Some of the tasks are already published. I think I'll keep it because we are very close to the release deadline so that we have similarity across all the tasks.

.Prerequisites

* You have deployed {ocp-product-title} 4.14 or later.
* You are logged in to the {ocp-product-title} web console as a user with the `cluster-admin` role.
* You have completed the premigration checklists.
* You have the {SMProduct} {SMv2Version} Operator installed.
* You have the {SMProduct} 3 Operator installed.
* You created an `IstioCNI` resource.
* You have the `istioctl` tool installed.

> * You are running a `ClusterWide` `ServiceMeshControlPlane`.

Author: done

* You have installed the `bookinfo` application.

.Procedure

. Identify the namespaces that contain a 2.6 control plane by running the following command:
+
[source,terminal]
----
$ oc get smcp -A
----
+
.Sample output:
[source,terminal]
----
NAMESPACE      NAME                   READY   STATUS            PROFILES      VERSION   AGE
istio-system   install-istio-system   6/6     ComponentsReady   ["default"]   2.6.6     115m
----

. Create a YAML file named `ossm-3.yaml` that creates the {istio} resource for the 3.0 installation in the same namespace as the `ServiceMeshControlPlane` resource for the 2.6 installation.
+
.Example `Istio` resource
[source,yaml,subs="attributes,verbatim"]
----
apiVersion: sailoperator.io/v1alpha1
kind: Istio
metadata:
  name: ossm-3 <1>
spec:
  updateStrategy:
    type: RevisionBased
  namespace: istio-system <2>
  version: v1.24.3
  values:  
    meshConfig:

>     extensionProviders: <3>
>       - name: prometheus
>         prometheus: {}
>       - name: otel
>         opentelemetry:
>           port: 4317
>           service: otel-collector.opentelemetrycollector-3.svc.cluster.local

Reviewer 4: ident is slightly off
```suggestion
      extensionProviders: <3>
        - name: prometheus
          prometheus: {}
        - name: otel
          opentelemetry:
            port: 4317
            service: otel-collector.opentelemetrycollector-3.svc.cluster.local
```

----
<1> The `name`, `updateStrategy` and `version` fields are significant for injection labels.
<2> The 3.0 and 2.6 control planes must run in the same namespace.
<3> Optional: If you are migrating metrics and tracing, update the `extensionProviders` fields according to your tracing and metrics configurations.
+

> This example configuration does not use any `discoverySelectors`, which means the {istio} control plane has access to all namespaces. If you want to define `discoverySelectors` all data plane namespaces that you plan to migrate from 2.6 must be matched.

Author: Done.

Moved as note above the YAML code block to match structure of the related procedure.

+
[NOTE]
====

> To prevent the {SMProduct} 3.0 control plane from injecting proxies in the namespaces that have the `istio-injection=enabled` label applied and are still managed by {SMProduct} 2.6 control plane, do not use use the `default` name for the {istio} resource, and do not create the `default` revision tag at this time. You will create the `default` revision tag later in this procedure.

Author: An important part of what we're trying to communicate is that the you create the label later (not now).  I made slight edit here though.

> Namespaces that have `istio-injection=enabled` label applied and are managed by {SMProduct} 2.6 control plane. To prevent the {SMProduct} 3.0 control pane from injecting proxies in workloads, do not use use the `default` name for the {istio} resource, and do not create the `default` revision tag at this time. You will create the `default` revision tag later in this procedure.

Author: done

====

. Apply the YAML file by running the following command:
+
[source,terminal]
----
$ oc apply -f ossm-3.yaml
----

. Verify that the new `istiod` resource uses the existing root certificate by running the following command:

> +

Author: Done
